File: NotePal/Windows/NotePal.Windows/Services/TodoStore.cs
using NotePalWindows.Models;

namespace NotePalWindows.Services;

public sealed class TodoStore
{
    private readonly LocalDataStore _dataStore;

    public TodoStore(LocalDataStore dataStore)
    {
        _dataStore = dataStore;
    }

    public event EventHandler? Changed;
    public event EventHandler<TodoItem>? TodoCompleted;
    public Action<string>? ErrorRaised { get; set; }
    public string? LastError { get; private set; }
    public IReadOnlyList<TodoItem> Todos => _dataStore.Data.Todos
        .Where(todo => !todo.IsWellnessTodo)
        .OrderBy(todo => todo.IsCompleted)
        .ThenBy(todo => todo.DueDate ?? DateTimeOffset.MaxValue)
        .ThenByDescending(todo => todo.UpdatedAt)
        .ToList();
    public IReadOnlySet<string> RemindedTodoRevisions => _dataStore.Data.RemindedTodoRevisions;
    public int IncompleteCount => _dataStore.Data.Todos.Count(todo => !todo.IsCompleted && !todo.IsWellnessTodo);

    public TodoItem? CreateTodo()
    {
        try
        {
            var todo = new TodoItem();
            _dataStore.Update(data => data.Todos.Insert(0, todo));
            ClearError();
            return todo;
        }
        catch (Exception ex)
        {
            SetError($"无法保存待办：{ex.Message}");
            return null;
        }
    }

    public void UpdateTodo(TodoItem todo)
    {
        try
        {
            var existing = _dataStore.Data.Todos.FirstOrDefault(item => item.Id == todo.Id);
            var wasCompleted = existing?.IsCompleted == true;
            var updated = todo.Clone();
            updated.Title = string.IsNullOrWhiteSpace(updated.Title) ? "新待办" : updated.Title.Trim();
            updated.Description = string.IsNullOrWhiteSpace(updated.Description) ? null : updated.Description;
            updated.UpdatedAt = DateTimeOffset.Now;
            updated.WellnessKind = null;
            updated.WellnessReminderIntervalMinutes = null;

            _dataStore.Update(data =>
            {
                var index = data.Todos.FindIndex(item => item.Id == updated.Id);
                if (index >= 0)
                {
                    data.Todos[index] = updated;
                }
            });

            ClearError();
            if (!wasCompleted && updated.IsCompleted)
            {
                TodoCompleted?.Invoke(this, updated);
            }
        }
        catch (Exception ex)
        {
            SetError($"无法保存待办：{ex.Message}");
        }
    }

    public void DeleteTodo(Guid id)
    {
        try
        {
            _dataStore.Update(data => data.Todos.RemoveAll(todo => todo.Id == id));
            ClearError();
        }
        catch (Exception ex)
        {
            SetError($"无法删除待办：{ex.Message}");
        }
    }

    public IReadOnlyList<TodoItem> DueTodosNeedingReminder(DateTimeOffset asOf)
    {
        return _dataStore.Data.Todos
            .Where(todo => !todo.IsWellnessTodo
                && todo.IsDue(asOf)
                && !_dataStore.Data.RemindedTodoRevisions.Contains(todo.ReminderRevisionKey))
            .ToList();
    }

    public DateTimeOffset? NextUnremindedDueDate(DateTimeOffset asOf)
    {
        return _dataStore.Data.Todos
            .Where(todo => !todo.IsWellnessTodo
                && todo.DueDate is not null
                && !todo.IsCompleted
                && todo.DueDate > asOf
                && !_dataStore.Data.RemindedTodoRevisions.Contains(todo.ReminderRevisionKey))
            .Select(todo => (DateTimeOffset?)todo.DueDate!.Value)
            .Min();
    }

    public void MarkReminded(IEnumerable<TodoItem> todos)
    {
        try
        {
            var revisions = todos
                .Where(todo => !todo.IsWellnessTodo)
                .Select(todo => todo.ReminderRevisionKey)
                .ToList();

            if (revisions.Count == 0)
            {
                return;
            }

            _dataStore.Update(data =>
            {
                foreach (var revision in revisions)
                {
                    data.RemindedTodoRevisions.Add(revision);
                }
            });
            ClearError();
        }
        catch (Exception ex)
        {
            SetError($"无法保存提醒状态：{ex.Message}");
        }
    }

    private void ClearError()
    {
        LastError = null;
        Changed?.Invoke(this, EventArgs.Empty);
    }

    private void SetError(string message)
    {
        LastError = message;
        ErrorRaised?.Invoke(message);
        Changed?.Invoke(this, EventArgs.Empty);
    }
}

File: NotePal/Sources/NotePal/Stores/TodoStore.swift
import Foundation

@MainActor
final class TodoStore: ObservableObject {
    @Published private(set) var todos: [TodoItem] = []
    @Published private(set) var remindedTodoRevisions: Set<String> = []
    @Published var lastError: String?

    var onTodosChanged: (() -> Void)?
    var onTodoCompleted: ((TodoItem) -> Void)?
    var onError: ((String) -> Void)?

    var incompleteCount: Int {
        todos.filter { !$0.isCompleted && !$0.isWellnessTodo }.count
    }

    private let storage: LocalDataStore

    init(storage: LocalDataStore = .shared) {
        self.storage = storage
        reload()
    }

    func reload() {
        do {
            let data = try storage.update { data in
                data.todos.removeAll { $0.isWellnessTodo }
                cleanReminderState(in: &data)
            }
            todos = sort(data.todos.filter { !$0.isWellnessTodo })
            remindedTodoRevisions = data.remindedTodoRevisions
            lastError = nil
            onTodosChanged?()
        } catch {
            lastError = "无法加载待办：\(error.localizedDescription)"
            onError?(lastError ?? "无法加载待办。")
        }
    }

    @discardableResult
    func createTodo(
        title: String = "新待办",
        description: String? = nil,
        dueDate: Date? = nil
    ) -> TodoItem {
        let todo = TodoItem(title: title, description: description, dueDate: dueDate)

        persist { data in
            data.todos.insert(todo, at: 0)
            cleanReminderState(in: &data)
        }

        return todo
    }

    func updateTodo(_ todo: TodoItem) {
        var updated = todo
        updated.updatedAt = Date()
        let completedTodo = todos.first(where: { $0.id == updated.id && !$0.isCompleted && updated.isCompleted })

        persist { data in
            guard let index = data.todos.firstIndex(where: { $0.id == updated.id }) else {
                return
            }

            data.todos[index] = updated
            cleanReminderState(in: &data)
        }

        if completedTodo != nil {
            onTodoCompleted?(updated)
        }
    }

    func toggleCompleted(id: UUID) {
        var completedTodo: TodoItem?

        persist { data in
            guard let index = data.todos.firstIndex(where: { $0.id == id }) else {
                return
            }

            data.todos[index].isCompleted.toggle()
            data.todos[index].updatedAt = Date()
            if data.todos[index].isCompleted {
                completedTodo = data.todos[index]
            }
            cleanReminderState(in: &data)
        }

        if let completedTodo {
            onTodoCompleted?(completedTodo)
        }
    }

    func deleteTodo(id: UUID) {
        persist { data in
            data.todos.removeAll { $0.id == id }
            cleanReminderState(in: &data)
        }
    }

    func dueTodosNeedingReminder(asOf date: Date = Date()) -> [TodoItem] {
        todos.filter { todo in
            !todo.isWellnessTodo
                && todo.isDue(asOf: date)
                && !remindedTodoRevisions.contains(todo.reminderRevisionKey)
        }
    }

    func markReminded(_ dueTodos: [TodoItem]) {
        guard !dueTodos.isEmpty else {
            return
        }

        let keys = Set(dueTodos.filter { !$0.isWellnessTodo }.map(\.reminderRevisionKey))

        persist { data in
            if !keys.isEmpty {
                data.remindedTodoRevisions.formUnion(keys)
            }

            cleanReminderState(in: &data)
        }
    }

    private func persist(_ mutation: (inout NotePalData) -> Void) {
        do {
            let data = try storage.update(mutation)
            todos = sort(data.todos.filter { !$0.isWellnessTodo })
            remindedTodoRevisions = data.remindedTodoRevisions
            lastError = nil
            onTodosChanged?()
        } catch {
            lastError = "无法保存待办：\(error.localizedDescription)"
            onError?(lastError ?? "无法保存待办。")
        }
    }

    private func sort(_ todos: [TodoItem]) -> [TodoItem] {
        todos.sorted { first, second in
            switch (first.dueDate, second.dueDate) {
            case let (firstDate?, secondDate?):
                return firstDate < secondDate
            case (.some, .none):
                return true
            case (.none, .some):
                return false
            case (.none, .none):
                return first.updatedAt > second.updatedAt
            }
        }
    }
}

private func cleanReminderState(in data: inout NotePalData) {
    let currentKeys = Set(data.todos.filter { !$0.isWellnessTodo }.map(\.reminderRevisionKey))
    data.remindedTodoRevisions = data.remindedTodoRevisions.intersection(currentKeys)
}
